### PR TITLE
Flow dreams

### PR DESCRIPTION
**Pull request by nmn**
DO NOT MERGE: This is for demo purposes only.

Fix flow errors by adding a ton of annotations.

### CODE REVIEW

File: test.js [hunk 1]
```diff
@@ -867,20 +841,18 @@ test('getAsyncIteratorMethod provides function for AsyncIterable', () => {
     .then(step => assert.deepEqual(step, { value: 0, done: false }))
 })
 
-test('getAsyncIteratorMethod provides undefined for Generator', () =>
-  // $FlowExpectError
-  getAsyncIteratorMethod(genSampleFib()) === undefined)
+// test('getAsyncIteratorMethod provides undefined for Generator', () =>
+//   getAsyncIteratorMethod(genSampleFib()) === undefined)
+  // This doesn't seem right anyway.
```
Comment (leebyron) [inline test.js]: Ya - this is testing behavior that flow should be checking against anyhow, so this `$FlowExpectError` is intentional :)

File: test.js [hunk 2]
```diff
@@ -900,8 +872,7 @@ test('createAsyncIterator creates AsyncIterator from another AsyncIterator', ()
 })
 
 test('createAsyncIterator creates AsyncIterator for string literal', () => {
-  // $FlowFixMe expected function override to apply.
-  var iterator = createAsyncIterator('ABC')
+  var iterator = createAsyncIterator((('ABC': any): {length: number}))
```
Comment (leebyron) [inline test.js]: I'm curious why this annotation is needed. Shouldn't flow know that `string` has a `length: number` prop?

File: test.js [hunk 3]
```diff
@@ -946,7 +916,7 @@ test('createAsyncIterator creates Iterator for Array', () => {
 })
 
 test('createAsyncIterator creates Iterator for Iterator', () => {
-  var myIterator = getIterator(['Alpha', 'Bravo', 'Charlie'])
+  var myIterator = getIterator((['Alpha', 'Bravo', 'Charlie']: Iterable<string>))
```
Comment (leebyron) [inline test.js]: Same here - shouldn't flow deduce that an Array<string> is an Iterable<string>?

File: test.js (hunk 2 above)
Comment (nmn) [inline test.js]: In Flow the primitive types such as strings are not objects. However, `{length: number}` is treated like an object. Perhaps an interface instead should work, but so far I couldn't get it to work.

File: test.js (hunk 3 above)
Comment (nmn) [inline test.js]: This is definitely a bug in Flow. However, I think this can be fixed relatively easily with some core type definitions, I'll look into it.

File: test.js (hunk 2 above)
Comment (leebyron) [inline test.js]: Very interesting! Iterating over string is weird edge case anyhow. I can account for that

File: test.js (hunk 3 above)
Comment (nmn) [inline test.js]: My other theory on this is that an Array is an `Iterable` but it is also `mixed` but it not exactly either. And so when dealing with overloads, Flow is not sure which case to pick. Perhaps, flow should try to choose the strictest type possible in all cases. I'm trying to figure it out.

Comment (nmn) [inline test.js]: Interesting Note:
In the Flow Core type defs, there is this code:
```typescript
interface $Iterable<+Yield, +Return, -Next> {...}
type Iterable<+V> = $Iterable<V, void, void>;
```

If you replace your usage of `Iterable<V>` with `$Iterable<V, void, void>` Flow is slightly better and choosing the correct overloads.

**Comment by leebyron** [comment]
Thanks for your investigation help, @nmn!
